lib/experimental/auto-inserting-blocks.php

```diff
@@ -0,0 +1,246 @@
+			continue;
+		}
+		$blocks                  = parse_blocks( $block_template->content );
+		$block_template->content = gutenberg_serialize_blocks( $blocks );
+	}
+
+	return $query_result;
+}
+add_filter( 'get_block_templates', 'gutenberg_parse_and_serialize_block_templates', 10, 1 );
+
+/**
+ * Filters the block template object after it has been (potentially) fetched from the theme file.
+ *
+ * By parsing a block template's content and then reserializing it
+ * via `gutenberg_serialize_blocks()`, we are able to run filters
+ * on the parsed blocks.
+ *
+ * @param WP_Block_Template|null $block_template The found block template, or null if there is none.
+ */
+function gutenberg_parse_and_serialize_blocks( $block_template ) {
+
+	$blocks                  = parse_blocks( $block_template->content );
+	$block_template->content = gutenberg_serialize_blocks( $blocks );
+
+	return $block_template;
+}
+add_filter( 'get_block_file_template', 'gutenberg_parse_and_serialize_blocks', 10, 1 );
+
+// Helper functions.
+// -----------------
+// The sole purpose of the following two functions (`gutenberg_serialize_block`
+// and `gutenberg_serialize_blocks`), which are otherwise copies of their unprefixed
+// counterparts (`serialize_block` and `serialize_blocks`) is to apply a filter
+// (also called `gutenberg_serialize_block`) as an entry point for modifications
+// to the parsed blocks.
+
+/**
+ * Filterable version of `serialize_block()`.
+ *
+ * This function is identical to `serialize_block()`, except that it applies
+ * the `gutenberg_serialize_block` filter to each block before it is serialized.
+ *
+ * @param array $block The block to be serialized.
+ * @return string The serialized block.
+ *
+ * @see serialize_block()
+ */
+function gutenberg_serialize_block( $block ) {
+	$block_content = '';
+
+	/**
+	 * Filters a parsed block before it is serialized.
+	 *
+	 * @param array $block The block to be serialized.
+	 */
+	$block = apply_filters( 'gutenberg_serialize_block', $block );
+
+	$index = 0;
+	foreach ( $block['innerContent'] as $chunk ) {
+		if ( is_string( $chunk ) ) {
+			$block_content .= $chunk;
+		} else { // Compare to WP_Block::render().
+			$inner_block    = $block['innerBlocks'][ $index++ ];
+			$block_content .= gutenberg_serialize_block( $inner_block );
+		}
+	}
+
+	if ( ! is_array( $block['attrs'] ) ) {
+		$block['attrs'] = array();
+	}
+
+	return get_comment_delimited_block_content(
+		$block['blockName'],
+		$block['attrs'],
+		$block_content
+	);
+}
+
+/**
+ * Filterable version of `serialize_blocks()`.
+ *
+ * This function is identical to `serialize_blocks()`, except that it applies
+ * the `gutenberg_serialize_block` filter to each block before it is serialized.
+ *
+ * @param array $blocks The blocks to be serialized.
+ * @return string[] The serialized blocks.
+ *
+ * @see serialize_blocks()
+ */
+function gutenberg_serialize_blocks( $blocks ) {
+	return implode( '', array_map( 'gutenberg_serialize_block', $blocks ) );
```

Comment:
as much fun as it is, there could be reason here to avoid the array abstractions since they are so relatively bad performance wise in PHP.

```php
$output = '';
foreach ( $blocks as $block ) {
	$output .= gutenberg_serialize_block( $block );
}
return $output;
```

will this be measurable? I doubt it, but it's characteristics are significantly better than `implode( array_map( ) )` and it's not that different in terms of PHP clarity. just a thought.

Reply:
This is just copied from Core's `serialize_blocks` though: https://github.com/WordPress/wordpress-develop/blob/4569aa17ae348c049bbacf8af8c6497a2e673b75/src/wp-includes/blocks.php#L820-L822.

I'd rather keep this part as consistent as possible with Core TBH 😬 (especially so we don't have anything that distracts from the _functional_ difference, i.e. the added filter in `gutenberg_serialize_block()`).